File: 2022/19.cpp
#include <array>
#include <cstring>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <string_view>
#include <vector>

struct Blueprint {
    int ore_ore;
    int clay_ore;
    int obsidian_ore;
    int obsidian_clay;
    int geode_ore;
    int geode_obsidian;
};

std::ostream& operator<<(std::ostream& o, Blueprint const& bp)
{
    o << "[\n"
      << "  ore_robot: " << bp.ore_ore << " ore\n"
      << "  clay_robot: " << bp.clay_ore << " ore\n"
      << "  obsidian_robot: " << bp.obsidian_ore << " ore, " << bp.obsidian_clay << " clay\n"
      << "  geode_robot: " << bp.geode_ore << " ore, " << bp.geode_obsidian << " obsidian\n"
      << "]";
    return o;
}

std::vector<Blueprint> readBlueprints(std::ifstream& input)
{
    std::vector<Blueprint> blueprints;
    std::array<char, 200> lineBuf;

    while (!input.eof()) {
        input.getline(&lineBuf[0], 200);
        std::string line(&lineBuf[0], std::strlen(&lineBuf[0]));
        if (line.size() == 0) {
            continue;
        }

        auto reader = std::stringstream{line};
        std::string dummy;
        // Blueprint 30: Each ore robot costs 4 ore. Each clay robot costs 4 ore.
        // Each obsidian robot costs 4 ore and 12 clay. Each geode robot costs
        // 4 ore and 19 obsidian.
        int ore_ore;
        int clay_ore;
        int obsidian_ore;
        int obsidian_clay;
        int geode_ore;
        int geode_obsidian;
        reader >> dummy >> dummy >> dummy >> dummy >> dummy >> dummy >> ore_ore
               >> dummy >> dummy >> dummy >> dummy >> dummy >> clay_ore
               >> dummy >> dummy >> dummy >> dummy >> dummy >> obsidian_ore
               >> dummy >> dummy >> obsidian_clay
               >> dummy >> dummy >> dummy >> dummy >> dummy >> geode_ore
               >> dummy >> dummy >> geode_obsidian;
        blueprints.push_back(Blueprint{
            ore_ore, clay_ore, obsidian_ore, obsidian_clay, geode_ore, geode_obsidian});
    }

    return blueprints;
}

struct State {
    int ore;
    int clay;
    int obsidian;
    int geode;
    int ore_robot;
    int clay_robot;
    int obsidian_robot;
    int geode_robot;
};

enum NewRobot {
    NONE = 0,
    ORE = 1,
    CLAY = 2,
    OBSIDIAN = 3,
    GEODE = 4
};

void add_state(NewRobot which, State const& state, Blueprint const& bp, std::vector<State>& list)
{
    int add_ore = 0;
    int add_clay = 0;
    int add_obsidian = 0;
    int add_geode = 0;
    int cost_ore = 0;
    int cost_clay = 0;
    int cost_obsidian = 0;
    switch (which) {
    case NewRobot::ORE:
        add_ore = 1;
        cost_ore = bp.ore_ore;
        break;
    case NewRobot::CLAY:
        add_clay = 1;
        cost_ore = bp.clay_ore;
        break;
    case NewRobot::OBSIDIAN:
        add_obsidian = 1;
        cost_ore = bp.obsidian_ore;
        cost_clay = bp.obsidian_clay;
        break;
    case NewRobot::GEODE:
        add_geode = 1;
        cost_ore = bp.geode_ore;
        cost_obsidian = bp.geode_obsidian;
        break;
    }
    list.push_back({
        state.ore + state.ore_robot - cost_ore,
        state.clay + state.clay_robot - cost_clay,
        state.obsidian + state.obsidian_robot - cost_obsidian,
        state.geode + state.geode_robot,
        state.ore_robot + add_ore,
        state.clay_robot + add_clay,
        state.obsidian_robot + add_obsidian,
        state.geode_robot + add_geode
    });
}


int geodesFromBlueprint(Blueprint const& bp, int const time)
{
    bool part2_pruning = (time == 32);
    int const clay_robot_limit = part2_pruning? 10 : 8;
    int const ore_robot_limit = part2_pruning? 10: 7;
    int const obsidian_robot_limit = part2_pruning? 10: 20;
    std::vector<State> current_states;
    std::vector<State> new_states;
    current_states.push_back({0, 0, 0, 0, 1, 0, 0, 0});
    int geode_robot_count = 0;
    int obsidian_robot_count = 0;
    for (int minute = 0; minute < time; ++minute) {
        //std::cout << "minute " << minute << ", states " << current_states.size() << std::endl;
        new_states.clear();
        for (auto const& state : current_states) {
            /* Unsafe optimization: If we already know a state with N geode robots where N is
             * more than 1 higher than the current state's geode robots, then prune that branch. */
            if (part2_pruning) {
                if (state.geode_robot > geode_robot_count) {
                    geode_robot_count = state.geode_robot;
                }
                if (state.geode_robot < geode_robot_count - 1) {
                    continue;
                }

                /* Follow-up unsafe optimization - do the same on obsidian robots */

                if (state.obsidian_robot > obsidian_robot_count) {
                    obsidian_robot_count = state.obsidian_robot;
                }
                if (state.obsidian_robot < obsidian_robot_count - 1) {
                    continue;
                }
            }

            /* No point in building clay or obsidian robots in the last 2 minutes */
            if ((minute >= time - 2 && state.clay_robot < 1)
                || (minute >= time - 1 && state.obsidian_robot < 1)) {
                continue;
            }

            /* Unsafe optimization: Build a geode robot whenever you can */
            if (state.ore >= bp.geode_ore && state.obsidian >= bp.geode_obsidian) {
                add_state(NewRobot::GEODE, state, bp, new_states);
                continue;
            }
            if (minute < time - 1) {
                /* Unsafe optimization: build an obsidian robot whenever you can */
                if (state.ore >= bp.obsidian_ore && state.clay >= bp.obsidian_clay && state.obsidian_robot < obsidian_robot_limit) {
                    add_state(NewRobot::OBSIDIAN, state, bp, new_states);
                    continue;
                }
                /* Don't build clay and ore robots in the last few minutes. Plus
                 * unsafe optimization: Don't build more of these robots when there
                 * are already N (constants above). */
                if (minute < time - 2) {
                    if (state.ore >= bp.clay_ore && state.clay_robot < clay_robot_limit) {
                        add_state(NewRobot::CLAY, state, bp, new_states);
                    }
                    if (minute < time - 3) {
                        if (state.ore >= bp.ore_ore && state.ore_robot < ore_robot_limit) {
                            add_state(NewRobot::ORE, state, bp, new_states);
                        }
                    }
                }
            }
            add_state(NewRobot::NONE, state, bp, new_states);
        }
        current_states.swap(new_states);
    }
    int best_geode = 0;
    for (auto const& state : current_states) {
        if (state.geode > best_geode) {
            best_geode = state.geode;
        }
    }
    return best_geode;
}


int main()
{
    auto input = std::ifstream{"input19.txt", std::ios::in};
    if (!input.is_open()) {
        std::cerr << "failed to open input file" << std::endl;
        return 1;
    }

    std::vector<Blueprint> testBlueprints = {
        {4, 2, 3, 14, 2, 7},
        {2, 3, 3, 8, 3, 12},
    };
    std::cout << testBlueprints.size() << " test blueprints" << std::endl;

    auto const blueprints = readBlueprints(input);
    std::cout << blueprints.size() << " blueprints" << std::endl;

    // part 1 test
    std::cout << "-- part 1 test --" << std::endl;
    for (int index = 0; index < testBlueprints.size(); ++index) {
        std::cout << "Blueprint " << (index + 1)
                  << ": " << geodesFromBlueprint(testBlueprints[index], 24) << " geodes" << std::endl;
    }

    std::cout << "-- part 1 real input --" << std::endl;
    int part1 = 0;
    for (int index = 0; index < blueprints.size(); ++index) {
        int const geodes = geodesFromBlueprint(blueprints[index], 24);
        part1 += (index + 1) * geodes;
        std::cout << "Blueprint " << (index + 1) << ": " << geodes << " geodes" << std::endl;
    }
    std::cout << "Part 1 total: " << part1 << std::endl;

    // part 2 test
    std::cout << "-- part 2 test --" << std::endl;
    for (int index = 0; index < testBlueprints.size(); ++index) {
        std::cout << "Blueprint " << (index + 1)
                  << ": " << geodesFromBlueprint(testBlueprints[index], 32) << " geodes" << std::endl;
    }

    // part 2 real
    std::cout << "-- part 2 real input --" << std::endl;
    int part2 = 1;
    for (int index = 0; index < 3; ++index) {
        int const geodes = geodesFromBlueprint(blueprints[index], 32);
        part2 *= geodes;
        std::cout << "Blueprint " << (index + 1) << ": " << geodes << " geodes" << std::endl;
    }

    std::cout << "Part 2 total: " << part2 << std::endl;

    return 0;
}

File: 2022/19.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text.RegularExpressions;

namespace AoC2022
{
    internal class Day19
    {
        public long totalStatesProbed = 0;

        internal enum Material
        {
            Ore = 0,
            Clay = 1,
            Obsidian = 2,
            Geode = 3,
        }

        internal class Materials
        {
            public int ore { get; private set; }
            public int clay { get; private set; }
            public int obsidian { get; private set; }
            public int geode { get; private set; }

            public Materials(int ore, int clay, int obsidian, int geode)
            {
                this.ore = ore;
                this.clay = clay;
                this.obsidian = obsidian;
                this.geode = geode;
            }

            public int this[Material i]
            {
                get
                {
                    return i switch
                    {
                        Material.Ore => ore,
                        Material.Clay => clay,
                        Material.Obsidian => obsidian,
                        Material.Geode => geode,
                        _ => throw new IndexOutOfRangeException(),
                    };
                }
                set
                {
                    switch (i)
                    {
                        case Material.Ore: ore = value; break;
                        case Material.Clay: clay = value; break;
                        case Material.Obsidian: obsidian = value; break;
                        case Material.Geode: geode = value; break;
                        default: throw new IndexOutOfRangeException();
                    }
                }
            }

            public Materials AddOne(Material material)
            {
                var result = new Materials(ore, clay, obsidian, geode);
                result[material] += 1;
                return result;
            }

            public override string ToString()
            {
                return $"[Or{ore}, Cl {clay}, Ob {obsidian}, Ge {geode}]";
            }
        }

        internal class Blueprint
        {
            public int index {get; private set;}

            public Materials ore_cost { get; private set; }
            public Materials clay_cost { get; private set; }
            public Materials obsidian_cost { get; private set; }
            public Materials geode_cost { get; private set; }

            public Blueprint(string line)
            {
                var numbers = Regex.Matches(line, @"\d+").Select(w => int.Parse(w.Value)).ToList();
                Debug.Assert(numbers.Count == 7);
                index = numbers[0];
                ore_cost = new Materials(numbers[1], 0, 0, 0);
                clay_cost = new Materials(numbers[2], 0, 0, 0);
                obsidian_cost = new Materials(numbers[3], numbers[4], 0, 0);
                geode_cost = new Materials(numbers[5], 0, numbers[6], 0);
            }

            public Materials this[Material type]
            {
                get
                {
                    return type switch
                    {
                        Material.Ore => ore_cost,
                        Material.Clay => clay_cost,
                        Material.Obsidian => obsidian_cost,
                        Material.Geode => geode_cost,
                        _ => throw new IndexOutOfRangeException(),
                    };
                }
            }

            // Return maximal amount of material of type 'type' to build any of the kinds of robot.
            public int maxMaterialCost(Material type)
            {
                var matArray = (Material[]) Enum.GetValues(typeof(Material));
                return matArray.Select(mat => this[mat][type]).Max();
            }

            public override string ToString()
            {
                return $"[Blueprint {index}. Robot costs: ore {ore_cost}, clay {clay_cost}, obsidian {obsidian_cost}, geode {geode_cost}]";
            }
        }

        private List<Blueprint> blueprints;

        public Day19(string input)
        {
            this.blueprints = input.Split("\n")
                                   .Where(line => line.Length > 0)
                                   .Select(line => new Blueprint(line))
                                   .ToList();
        }

        internal int part1()
        {
            const int TimePart1 = 24;
            return blueprints.Select(blueprint => evaluateBlueprint(blueprint, TimePart1) * blueprint.index)
                             .Sum();
        }

        internal int part2()
        {
            const int TimePart2 = 32;
            const int BlueprintCount = 3;
            return blueprints.Take(BlueprintCount)
                             .Select(blueprint => evaluateBlueprint(blueprint, TimePart2))
                             .Aggregate(1, (val, total) => total * val);
        }

        private class ProductionState {
            public int timeLeft { get; set; }
            public Materials resource { get; set; }
            public Materials robots { get; set; }

            public ProductionState(int time)
            {
                timeLeft = time;
                resource = new Materials(0, 0, 0, 0);
                robots = new Materials(1, 0, 0, 0);
            }

            public ProductionState(int time, Materials resource, Materials robots)
            {
                timeLeft = time;
                this.resource = resource;
                this.robots = robots;
            }
        }


        internal int evaluateBlueprint(Blueprint b, int time)
        {
            // We can use a queue or a stack for open states. It seems that using a queue reduces
            // the number of states that we need to visit to ~85% of those needed when using a stack,
            // at least on the given input. So we'll use the queue.
            var queue = new Queue<ProductionState>();
            queue.Enqueue(new ProductionState(time));

            int mostGeodes = 0;

            var maxRobotsNeeded = new Materials(
                b.maxMaterialCost(Material.Ore),
                b.maxMaterialCost(Material.Clay),
                b.maxMaterialCost(Material.Obsidian),
                b.maxMaterialCost(Material.Geode)
            );

            while (queue.Count > 0)
            {
                totalStatesProbed++;
                var state = queue.Dequeue();
                if (state.resource.geode > mostGeodes)
                {
                    mostGeodes = state.resource.geode;
                }
                if (state.timeLeft <= 1) {
                    // No time to build anything else; end state
                    continue;
                }

                // Prune states that are not perspective - a real time saver!
                // If we were to build a geode robot from now on in every minute, how much could
                // we possibly harvest?
                var potential = (state.timeLeft - 1) * (state.timeLeft - 1 + 1) / 2;
                if (state.resource.geode + potential <= mostGeodes)
                {
                    // No point continuing with this path; it cannot beat the current best
                    continue;
                }

                foreach (Material buildNext in Enum.GetValues(typeof(Material)))
                {
                    // Is it possible to build this material robot?
                    int timeToBuild = 0;
                    bool impossibleToBuild = false;
                    var cost = b[buildNext];
                    foreach (Material resource in Enum.GetValues(typeof(Material)))
                    {
                        if (cost[resource] > 0)
                        {
                            if (state.robots[resource] == 0)
                            {
                                // impossible to build next; we don't have the right dependency robot yet
                                impossibleToBuild = true;
                                break;
                            }
                            var resourceNeeded = Math.Max(cost[resource] - state.resource[resource], 0);
                            var timeNeeded = (resourceNeeded + state.robots[resource] - 1) / state.robots[resource]; // rounded up to int
                            timeToBuild = Math.Max(timeToBuild, timeNeeded);
                        }
                    }
                    // +1 for the time to actually build the robot
                    timeToBuild += 1;
                    // Time needs to be strictly greater, otherwise we'd just build the robot but not have time to use it
                    if (impossibleToBuild || timeToBuild > state.timeLeft)
                    {
                        // will not build this robot type next
                        continue;
                    }

                    // At this point, it is possible to build the robot of type 'buildNext', and it would take 'timeToBuild'.
                    // But is it useful to build that type? (It is always useful to build a geode robot)
                    if (buildNext != Material.Geode)
                    {
                        if (state.robots[buildNext] >= maxRobotsNeeded[buildNext])
                        {
                            // We already have enough robots of this type that we will ever need
                            continue;
                        }
                        // It doesn't make sense to build anything else than a geode robot if it's our last one
                        if (state.timeLeft <= 2)
                        {
                            continue;
                        }
                    }

                    // It is possible and might make sense to build this robot. Let's build it.

                    // How much of each material will be harvested while we're waiting to build the robot.
                    // Geodes are counted differently, because we're not consuming them. They are booked at
                    // the time when we complete building the robot
                    var resourcesLeft = new Materials(
                        state.resource.ore + state.robots.ore * timeToBuild - cost.ore,
                        state.resource.clay + state.robots.clay * timeToBuild - cost.clay,
                        state.resource.obsidian + state.robots.obsidian * timeToBuild - cost.obsidian,
                        state.resource.geode
                    );
                    var newRobots = state.robots.AddOne(buildNext);
                    var newTimeLeft = state.timeLeft - timeToBuild;

                    if (buildNext == Material.Geode)
                    {
                        resourcesLeft[Material.Geode] += newTimeLeft;
                    }

                    queue.Enqueue(new ProductionState(newTimeLeft, resourcesLeft, newRobots));
                }
            }

            Console.WriteLine($"Blueprint {b.index} -> most geodes {mostGeodes}");

            return mostGeodes;
        }
    }
}
